src/libtsduck/dtv/analysis/tsTr101Analyzer.cpp

```diff
@@ -0,0 +1,503 @@
+        const auto pmt = PMT(_duck, table);
+
+        // Ensure all PIDs are assigned to this service.
+        for (auto it = pmt.streams.begin(); it != pmt.streams.end(); it++) {
+            const auto s2 = getService(it->first);
+            s2->_type = ServiceContext::Assigned;
+            s2->_pmt_service_id = service->_pmt_service_id;
+        }
+
+        // Remove PIDs that are Assigned to this service ID.
+        for (auto it = _services.begin(); it != _services.end(); ++it) {
+            if (it->second->_type == ServiceContext::Assigned && it->second->_pmt_service_id == service->_pmt_service_id) {
+                if (!pmt.streams.contains(it->first)) {
+                    it->second->_type = ServiceContext::Unassigned;
+                }
+            }
+        }
+    }
+}
+
+void ts::TR101_290Analyzer::handleSection(SectionDemux& demux, const Section& table)
+{
+    if (table.sectionNumber() != 0) {
+        // we only care about the first packet that has arrived.
+        return;
+    }
+
+    auto service = getService(table.sourcePID());
+
+    if (table.sourcePID() == PID_PAT && table.tableId() != TID_PAT) {
+        service->pat_err.count++;
+        service->pat_err2.count++;
+    }
+
+    if (service->_type == ServiceContext::Pmt && table.tableId() != TID_PMT) {
+        service->pmt_err.count++;
+        service->pmt_err2.count++;
+    }
+
+    if (table.tableId() == TID_PAT && service->_type == ServiceContext::Pat) {
+        service->_type = ServiceContext::Pat;
+        if (service->_last_table_ts != INVALID_PCR) {
+            auto diff = long(_currentTimestamp - service->_last_table_ts);
+            service->pat_err.pushSysClockFreq(diff);
+            service->pat_err2.pushSysClockFreq(diff);
+            if (diff > PAT_INTERVAL) {
+                service->pat_err.count++;
+                service->pat_err2.count++;
+            }
+        }
+        service->_last_table_ts = _currentTimestamp;
+    }
+    else if (table.tableId() == TID_PMT && service->_type == ServiceContext::Pmt) {
+        if (service->_last_table_ts != INVALID_PCR) {
+            auto diff = long(_currentTimestamp - service->_last_table_ts);
+            service->pmt_err.pushSysClockFreq(diff);
+            service->pmt_err2.pushSysClockFreq(diff);
+            if (diff > PMT_INTERVAL) {
+                service->pmt_err.count++;
+                service->pmt_err2.count++;
+            }
+        }
+        service->_last_table_ts = _currentTimestamp;
+    }
+    else if (table.tableId() == TID_CAT) {
+        _lastCatIndex = _packetIndex;
+
+        for (auto & [_, s2] : _services) {
+            s2->cat_error = 0;
+        }
+    }
+}
+void ts::TR101_290Analyzer::handleInvalidSection(SectionDemux& demux, const DemuxedData& data) {
+    auto _service = getService(data.sourcePID());
+    // todo: it is not guaranteed that it is a CRC error that causes this issue.
+    _service->crc_error++;
```

Comment:
Am I bad in assuming that this will always be a CRC error? If not, how do I check if this was caused by a CRC error?

Reply:
No, there are many possible reasons for a section to be invalid.
See an example here:
https://github.com/tsduck/tsduck/blob/cc250477985e3e637ad03ae9f89bbdae6294655a/src/libtsduck/dtv/signalization/tsTablesLogger.cpp#L822

I recomputes the CRC to determine if this is a CRC error.

Reply:
Since the demux already did all checks before declaring a section invalid, it is more consistent to propagate that status to the invalid section handler, rather than letting the handler do the job again.

Therefore, I added an enum type for section status and the profile of the handler has been modified to add it as parameter:

https://github.com/tsduck/tsduck/blob/061b28307c5522d70859c34a287451eed9eba68e/src/libtsduck/dtv/demux/tsInvalidSectionHandlerInterface.h#L38

The status values are defined in class `Section`:

https://github.com/tsduck/tsduck/blob/061b28307c5522d70859c34a287451eed9eba68e/src/libtsduck/dtv/signalization/tsSection.h#L53

Reply:
Sweet! Added, thanks!